### PR TITLE
(FM-7625) implement finegrained configuration support for SSL verification

### PR DESCRIPTION
**Pull request by Thomas-Franklin**
This will make use of `verify_mode` control in order to do `peer`, `client-once`, or `fail-no-peer` verification.

Can also specify which `ca_file` to use for verification other wise it will use the certs in `OpenSSL::X509::DEFAULT_CERT_FILE`, e.g.:

 `ruby -ropenssl -e 'puts OpenSSL::X509::DEFAULT_CERT_FILE'`

The SHA256 fingerprint can be specified for additional verification.

### CODE REVIEW

**Comment by Thomas-Franklin** [comment]
Will be adding unit tests tomorrow. So this is more for early review at this point.

File: README.md [hunk 1]
```diff
@@ -86,6 +87,28 @@ This will sign the certificate and set up the device for Puppet.
...
+* `ssl.ca_file`: the full path to a CA certificate in PEM format. The certificate of the target needs to be signed by this CA. The file needs to exist on the proxy agent's local file system. Only one of `ssl.ca_file` and `ssl.fingerprint` may be provided. Example: `'/etc/ssl/certs/Go_Daddy_Root_Certificate_Authority_-_G2.pem'`
+* `ssl.fingerprint`: a string specifying the SHA256 fingerprint of the firewall's certificate in hex notation. This can be generated by `openssl x509 -sha256 -fingerprint -noout -in cert.pem` or seen in your browser's SSL certificate information. Only one of `ssl.ca_file` and `ssl.fingerprint` may be provided. Example: `'9A:6E:C0:12:E1:A7:DA:9D:BE:34:19:4D:47:8A:D7:C0:DB:18:22:FB:07:1D:F1:29:81:49:6E:D1:04:38:41:13'`
+* `ssl.ciphers`: array specifying the allowed ciphers for the connection.
+* `ssl.versions`: a string representing the ssl version, e.g. `TLSv1`
+
+*NOTE*: If you accept the possibility that an attacker can hijack your firewall configuration, you can turn of SSL verification completely by setting `ssl: off`.
```
Comment (DavidS) [inline README.md]: The same is true for `ssl.mode: off`. The distinction is `ssl:off` causes a connection to port 80 with http, while `ssl.mode:off` would connect to port 443 with https, but not verify the certificate (still allowing a man in the middle attack).

File: README.md [hunk 2]
```diff
@@ -86,6 +87,28 @@ This will sign the certificate and set up the device for Puppet.
...
+
+```
+address: 10.0.10.20
+username: admin
+password: admin
+ssl.ciphers: [ECDHE-ECDSA-AES128-GCM-SHA256, ECDHE-RSA-AES128-GCM-SHA256, ECDHE-ECDSA-AES256-GCM-SHA384, ECDHE-RSA-AES256-GCM-SHA384, AES256-GCM-SHA384]
+ssl.mode: peer
+ssl.version: SSLv23
+ssl.fingerprint: EA:21:E5:8F:13:98:73:DB:A6:25:0D:10:1A:08:57:55:34:B4:2C:A8:73:B9:CE:DC:96:4A:74:70:14:A0:7B:6D
```
Comment (DavidS) [inline README.md]: will this require quotes?

**Comment by Thomas-Franklin** [comment]
Tagging @clairecadman for the readme changes. @DavidS the changes proposed have been made I believe :-)

File: lib/puppet/util/network_device/panos/device.rb [hunk 3]
```diff
@@ -118,15 +118,74 @@ def initialize(credentials)
...
+    # @return OpenSSL::SSL verification mode
+    def handle_verify_mode(verify_mode)
+      case verify_mode
+      when 'on'
+        OpenSSL::SSL::VERIFY_PEER
+      when 'off'
+        Puppet.warning("SSL verification turned off in configuration for $#{Puppet[:certname]}")
+        OpenSSL::SSL::VERIFY_NONE
+      else
+        raise Puppet::ResourceError, "#{verify_mode} is not a valid mode, " \
```
Comment (DavidS) [inline lib/puppet/util/network_device/panos/device.rb]: Add quotes around `#{verify_mode}` to ensure the resulting message is easily readable. In the worst case `ssl.mode: This` would lead to `"This is not a valid mode,..."` making it exceptionally hard to understand what's going on.

File: README.md [hunk 4]
```diff
@@ -86,6 +87,30 @@ This will sign the certificate and set up the device for Puppet.
...
+* `ssl.ca_file`: the full path to a CA certificate in PEM format. The certificate of the target needs to be signed by this CA. The file needs to exist on the proxy agent's local file system. Only one of `ssl.ca_file` and `ssl.fingerprint` may be provided. Example: `'/etc/ssl/certs/Go_Daddy_Root_Certificate_Authority_-_G2.pem'`. Alternatively it will use the certs in `OpenSSL::X509::DEFAULT_CERT_FILE`, e.g.:
+
+`ruby -ropenssl -e 'puts OpenSSL::X509::DEFAULT_CERT_FILE'` if no `ssl.ca_file` is provided
+* `ssl.fingerprint`: a string specifying the SHA256 fingerprint of the firewall's certificate in hex notation. This can be generated by `openssl x509 -sha256 -fingerprint -noout -in cert.pem` or seen in your browser's SSL certificate information. Only one of `ssl.ca_file` and `ssl.fingerprint` may be provided. Example: `'9A:6E:C0:12:E1:A7:DA:9D:BE:34:19:4D:47:8A:D7:C0:DB:18:22:FB:07:1D:F1:29:81:49:6E:D1:04:38:41:13'`
+* `ssl.ciphers`: array specifying the allowed ciphers for the connection.
+* `ssl.version`: a string representing the ssl version, e.g. `TLSv1`
```
Comment (DavidS) [inline README.md]: Can you link to documentation on which values are acceptable for `ciphers`, and `version` ?

File: lib/puppet/util/network_device/panos/device.rb [hunk 5]
```diff
@@ -118,15 +118,74 @@ def initialize(credentials)
...
                   Net::HTTP.start(@host, @port,
                                   use_ssl: true,
-                                  verify_mode: OpenSSL::SSL::VERIFY_NONE)
+                                  verify_mode: handle_verify_mode(@ssl_verify),
+                                  ca_file: @ca_file,
+                                  ssl_version: @ssl_version,
+                                  ciphers: @ciphers ? @ciphers.join(':') : @ciphers,
+                                  verify_callback: ->(preverify_ok, cert_store) do
+                                                     verify_callback(preverify_ok, cert_store)
+                                                   end)
```
Comment (DavidS) [inline lib/puppet/util/network_device/panos/device.rb]: You might be able to directly convert the `verify_callback` method into a proc by using the syntax described in https://stackoverflow.com/a/4294660/4918

File: lib/puppet/util/network_device/panos/device.rb [hunk 6]
```diff
@@ -118,15 +118,74 @@ def initialize(credentials)
...
+      return false unless preverify_ok || @fingerprint
+
+      end_cert = cert_store.chain[0]
+
+      return true unless end_cert.to_der == cert_store.current_cert.to_der
+      @fingerprint ? same_cert_fingerprint?(end_cert) : true
+    end
+
+    def same_cert_fingerprint?(end_cert)
+      OpenSSL::Digest::SHA256.hexdigest(end_cert.to_der).scan(%r{..}).map { |s| s.upcase }.join(':') == @fingerprint
```
Comment (DavidS) [inline lib/puppet/util/network_device/panos/device.rb]: This method is security sensitive and this exposes it to extra scrutiny. This method is badly factored. There's a lot going on and external API accesses and data transformations are deeply entangled.

Try factoring out the following methods:

* `hexdigest_from_cert`
* `fingerprint_from_hexdigest`

Then you can test the former by having a known certificate and a known hexdigest for it, while the latter method becomes pure data in/data out testing.

File: lib/puppet/util/network_device/panos/device.rb [hunk 7]
```diff
@@ -118,15 +118,74 @@ def initialize(credentials)
...
+        Puppet.warning("SSL verification turned off in configuration for $#{Puppet[:certname]}")
+        OpenSSL::SSL::VERIFY_NONE
+      else
+        raise Puppet::ResourceError, "#{verify_mode} is not a valid mode, " \
+              'valid modes are: "off" for no client verification, ' \
+              'and "on" for validating the certificate'
+      end
+    end
+
+    def verify_callback(preverify_ok, cert_store)
```
Comment (DavidS) [inline lib/puppet/util/network_device/panos/device.rb]: Given the delicateness of the code and the general lack of documentation from the ruby side of this, please add a short statement how this method is expected to behave (that is, what does it mean to return `true` or `false`), and what the method is trying to accomplish (`"This SSL verification callback accepts certificates matching @fingerprint, if one is specified"`)

File: lib/puppet/util/network_device/panos/device.rb [hunk 8]
```diff
@@ -118,15 +118,74 @@ def initialize(credentials)
...
+      # if ssl_fingerprint is specified then do not
+      # depend on the pre-verification checks,
+      # if no ssl_fingerprint and the preverify_checks fail
+      # then certificate verification will fail overall
+      return false unless preverify_ok || @fingerprint
+
+      end_cert = cert_store.chain[0]
+
+      return true unless end_cert.to_der == cert_store.current_cert.to_der
```
Comment (DavidS) [inline lib/puppet/util/network_device/panos/device.rb]: If you copy code from stackoverflow (https://stackoverflow.com/a/22108461/4918), keep the comments and link to the original answer.

File: lib/puppet/util/network_device/panos/device.rb [hunk 9]
```diff
@@ -118,15 +118,74 @@ def initialize(credentials)
...
+                                  verify_callback: ->(preverify_ok, cert_store) do
+                                                     verify_callback(preverify_ok, cert_store)
+                                                   end)
+                # want to raise the Puppet::ResourceError message when raised, i.e.
+                # wrong verify mode has been specified
+                rescue Puppet::ResourceError => message
+                  raise message
+                rescue OpenSSL::SSL::SSLError
+                  raise Puppet::ResourceError, "Certificate Verification has failed, ensure the correct `ssl_ca_file` \
+or `ssl_fingerprint` is specified#{@ciphers ? ', and ensure the correct `ssl_ciphers` are specified and supported.' : '.'}"
```
Comment (DavidS) [inline lib/puppet/util/network_device/panos/device.rb]: The syntax in the config file is `ssl.`, not `ssl_`. Check the entire commit for other locations where you've made the same typo.

File: spec/spec_helper_acceptance.rb [hunk 10]
```diff
@@ -50,6 +50,7 @@ def debug_output?
 address: #{@hostname}
 username: #{ENV['PANOS_TEST_USER'] || 'admin'}
 password: #{ENV['PANOS_TEST_PASSWORD'] || 'admin'}
+ssl.verify: off
```
Comment (DavidS) [inline spec/spec_helper_acceptance.rb]: An "easy-ish" way to do good acceptance testing of your code is to write up a quick script that downloads the SSL certificate from the target device and configures validation here.

Comment (Thomas-Franklin) [reply spec/spec_helper_acceptance.rb]: Unfortunately there would be more work involved as configuring the SSL certificate correctly, as the generated one on the device I believe does not match the domain name and common name has issues which makes it an invalid cert, which leads to `Verify return code: 21 (unable to verify the first certificate)` which is why I had to manually configure the device in order to test these things. An alternative option would be to have a task which creates and sets the cert, but we can discuss this tomorrow :-)

File: README.md [hunk 11]
```diff
@@ -86,6 +87,30 @@ This will sign the certificate and set up the device for Puppet.
 
 See the [`puppet device` documentation](https://puppet.com/docs/puppet/5.5/puppet_device.html)
 
+### SSL Certificate Verification
+
+To configure SSL certificate verification, add the following `ssl` keys to your credentials file:
+
+* `ssl.verify`: the mode in which to use for validation, valid options are `off` where no certificate verification will happen but HTTPS will still be used, or `on` where certificate verification will occur. It will default to `on`.
+* `ssl.ca_file`: the full path to a CA certificate in PEM format. The certificate of the target needs to be signed by this CA. The file needs to exist on the proxy agent's local file system. Only one of `ssl.ca_file` and `ssl.fingerprint` may be provided. Example: `'/etc/ssl/certs/Go_Daddy_Root_Certificate_Authority_-_G2.pem'`. Alternatively it will use the certs in `OpenSSL::X509::DEFAULT_CERT_FILE`, e.g.:
```
Comment (clairecadman) [inline README.md]: What's PEM?

Switch "may" for "can", if that't what you mean. 

Keep consistent, you say "example:", and then on the line below "e.g.". Maybe just say, "for example", every time.

File: README.md [hunk 12]
```diff
@@ -86,6 +87,30 @@ This will sign the certificate and set up the device for Puppet.
 
 See the [`puppet device` documentation](https://puppet.com/docs/puppet/5.5/puppet_device.html)
 
+### SSL Certificate Verification
+
+To configure SSL certificate verification, add the following `ssl` keys to your credentials file:
+
+* `ssl.verify`: the mode in which to use for validation, valid options are `off` where no certificate verification will happen but HTTPS will still be used, or `on` where certificate verification will occur. It will default to `on`.
```
Comment (clairecadman) [inline README.md]: Omit "will", and a slight reword: 
"...the mode to use for validation. Valid options are `off`, where no certificate verification happens but HTTPS is still used, or `on`, where certificate verification occurs. It defaults to `on`"

File: README.md [hunk 13]
```diff
@@ -86,6 +87,30 @@ This will sign the certificate and set up the device for Puppet.
...
+* `ssl.ca_file`: the full path to a CA certificate in PEM format. The certificate of the target needs to be signed by this CA. The file needs to exist on the proxy agent's local file system. Only one of `ssl.ca_file` and `ssl.fingerprint` may be provided. Example: `'/etc/ssl/certs/Go_Daddy_Root_Certificate_Authority_-_G2.pem'`. Alternatively it will use the certs in `OpenSSL::X509::DEFAULT_CERT_FILE`, e.g.:
+
+`ruby -ropenssl -e 'puts OpenSSL::X509::DEFAULT_CERT_FILE'` if no `ssl.ca_file` is provided
+* `ssl.fingerprint`: a string specifying the SHA256 fingerprint of the firewall's certificate in hex notation. This can be generated by `openssl x509 -sha256 -fingerprint -noout -in cert.pem` or seen in your browser's SSL certificate information. Only one of `ssl.ca_file` and `ssl.fingerprint` may be provided. Example: `'9A:6E:C0:12:E1:A7:DA:9D:BE:34:19:4D:47:8A:D7:C0:DB:18:22:FB:07:1D:F1:29:81:49:6E:D1:04:38:41:13'`
+* `ssl.ciphers`: array specifying the allowed ciphers for the connection.
```
Comment (clairecadman) [inline README.md]: *can* be provided.

File: README.md [hunk 14]
```diff
@@ -86,6 +87,30 @@ This will sign the certificate and set up the device for Puppet.
...
+* `ssl.version`: a string representing the ssl version, e.g. `TLSv1`
+
+*NOTE*: If you accept the possibility that an attacker can hijack your firewall configuration, you can turn of SSL verification completely by setting `ssl.verify: off`.
```
Comment (clairecadman) [inline README.md]: Maybe just "for security reasons, turn of SSL verification completely by setting `ssl.verify: off`" ?

File: README.md (hunk 11 above)
Comment (Thomas-Franklin) [reply README.md]: PEM is a file format for storing/sending cryptographic keys.

File: README.md (hunk 14 above)
Comment (Thomas-Franklin) [reply README.md]: Turning off SSL is not advised, which is why we need to highlight it is up to them if they want to turn it off.

File: lib/puppet/util/network_device/panos/device.rb (hunk 5 above)
Comment (Thomas-Franklin) [reply lib/puppet/util/network_device/panos/device.rb]: I did want to split it out to make it easier to test the method

File: README.md (hunk 14 above)
Comment (clairecadman) [reply README.md]: Maybe rephrase "If you accept the possibility that an attacker can hijack your firewall configuration" -  I'm not quite sure what this means. Also maybe make it clearer turning off SSL is not advised too.

Comment (Thomas-Franklin) [reply README.md]: Maybe, 

```
Although not advisable, SSL can be turned off by setting `ssl.verify: off`, in doing so you increase the risk of your firewall configuration being hijacked by a potential attacker.
```

Comment (clairecadman) [reply README.md]: Maybe "... you can turn off SSL by..." (active voice). And then, new sentence "In doing so, you..". But otherwise looks good!

Comment (willmeek) [reply README.md]: `*NOTE*: Although not advisable, you can turn off SSL by setting `ssl.verify: off`. In doing so you increase the risk of your firewall configuration being hijacked by a potential attacker.`

File: spec/spec_helper_acceptance.rb (hunk 10 above)
Comment (DavidS) [reply spec/spec_helper_acceptance.rb]: We'll have to live with this for now, then

File: lib/puppet/util/network_device/panos/device.rb (hunk 5 above)
Comment (DavidS) [reply lib/puppet/util/network_device/panos/device.rb]: :+1:

File: lib/puppet/util/network_device/panos/device.rb (hunk 6 above)
Comment (DavidS) [reply lib/puppet/util/network_device/panos/device.rb]: not quite what I was angling at, but good enough for now.
